File: modules/student-experience/src/main/java/com/campusos/portal/service/StaffScope.java
package com.campusos.portal.service;

import com.campusos.portal.domain.Actor;
import com.campusos.portal.domain.StaffRole;
import com.campusos.portal.domain.Student;
import com.campusos.portal.domain.TeachingAssignment;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

/**
 * THE STAFF EQUIVALENT OF Scope. Built only from the authenticated principal — nothing in it
 * is ever read from a request parameter, which is what retires the client-supplied actor.
 *
 * Two different breadths live here, deliberately:
 *
 *   canSee(student)  — REQUEST breadth. Widens with role: a class for FACULTY, a department
 *                      for HOD, the tenant for INSTITUTION/OFFICE.
 *   teaches(class,subject) — ACADEMIC-WRITE breadth. Teaching assignment ONLY, whatever roles
 *                      are held. An HOD who teaches nothing may author nothing.
 */
public record StaffScope(String tenantId, String staffId, String name, String department,
                         Set<StaffRole> roles, List<TeachingAssignment> assignments) {

    public StaffScope {
        if (tenantId == null || tenantId.isBlank())
            throw new IllegalStateException("tenantId missing — staff query refused");
        if (staffId == null || staffId.isBlank())
            throw new IllegalStateException("staffId missing — staff query refused");
        roles = roles == null ? Set.of() : Set.copyOf(roles);
        assignments = assignments == null ? List.of() : List.copyOf(assignments);
    }

    /* ------------------------------- roles -> actors ------------------------------- */

    public boolean hasRole(StaffRole r) { return roles.contains(r); }

    /** The Actors this principal may ever present to the engine. Never STUDENT, never SYSTEM. */
    public Set<Actor> actors() {
        Set<Actor> out = new LinkedHashSet<>();
        for (StaffRole r : roles) out.add(r.actor());
        return out;
    }

    public boolean mayAct(Actor a) { return actors().contains(a); }

    /* ------------------------------- request breadth ------------------------------- */

    /**
     * REQUEST breadth. Tenant equality is a precondition of every branch — there is no path
     * through this method that returns true for a student of another tenant.
     */
    public boolean canSee(Student s) {
        if (s == null || !tenantId.equals(s.tenantId)) return false;
        if (hasRole(StaffRole.INSTITUTION) || hasRole(StaffRole.OFFICE)) return true;
        if (hasRole(StaffRole.HOD) && department != null && department.equals(s.department)) return true;
        if (hasRole(StaffRole.FACULTY)) {
            for (TeachingAssignment a : assignments) {
                if (ClassKey.of(a).matches(s)) return true;
            }
        }
        return false;
    }

    /* ---------------------------- academic-write breadth ---------------------------- */

    /** True iff this staff member personally teaches this subject to this class. */
    public boolean teaches(ClassKey c, String subjectCode) {
        for (TeachingAssignment a : assignments) {
            if (ClassKey.of(a).equals(c) && a.subjectCode.equals(subjectCode)) return true;
        }
        return false;
    }

    /** True iff this staff member teaches at least one subject to this class. */
    public boolean teaches(ClassKey c) {
        for (TeachingAssignment a : assignments) {
            if (ClassKey.of(a).equals(c)) return true;
        }
        return false;
    }

    /** The classes this staff member teaches — the ONLY values the class picker is built from. */
    public List<ClassKey> classes() {
        LinkedHashSet<ClassKey> out = new LinkedHashSet<>();
        for (TeachingAssignment a : assignments) out.add(ClassKey.of(a));
        return List.copyOf(out);
    }

    public List<TeachingAssignment> subjectsIn(ClassKey c) {
        return assignments.stream().filter(a -> ClassKey.of(a).equals(c)).toList();
    }

    public boolean authors() { return !assignments.isEmpty(); }
}

File: modules/student-experience/src/service/StaffScope.js
import { StaffRole } from '../domain/enums.js';
import { ClassKey } from './ClassKey.js';

/**
 * THE STAFF EQUIVALENT OF Scope. Built only from the authenticated principal — nothing in it
 * is ever read from a request parameter, which is what retires the client-supplied actor.
 *
 * Two different breadths live here:
 *   canSee(student)  — REQUEST breadth. Widens with role.
 *   teaches(class, subject) — ACADEMIC-WRITE breadth. Teaching assignment ONLY.
 */
export class StaffScope {
  constructor(tenantId, staffId, name, department, roles, assignments) {
    if (!tenantId || !tenantId.trim()) throw new Error('tenantId missing — staff query refused');
    if (!staffId || !staffId.trim()) throw new Error('staffId missing — staff query refused');
    this.tenantId = tenantId;
    this.staffId = staffId;
    this.name = name;
    this.department = department;
    this.roles = new Set(roles ?? []);
    this.assignments = assignments ?? [];
  }

  hasRole(r) { return this.roles.has(r); }

  /** The Actors this principal may ever present to the engine. Never STUDENT, never SYSTEM. */
  actors() { return [...this.roles].map((r) => StaffRole.actor(r)); }

  mayAct(a) { return this.actors().includes(a); }

  /**
   * Statutory committees required by UGC regulations. Deliberately NOT folded into canSee()/
   * roster(): those two govern the FULL student profile (attendance, marks, every other
   * request) and widening them institution-wide for, say, an Anti-Ragging Committee seat would
   * hand that committee everyone's academic record, not just the ragging complaints that are
   * actually theirs to see. Committee reach is instead a narrow, grievance-only, category-
   * gated path — see StaffScopeResolver.grievanceRoster() and GrievanceVisibility.js.
   */
  static COMMITTEE_ROLES = [
    StaffRole.ICC, StaffRole.ANTI_RAGGING, StaffRole.SC_ST_CELL,
    StaffRole.EQUAL_OPPORTUNITY_CELL, StaffRole.RTI_OFFICER, StaffRole.OMBUDSPERSON,
  ];

  isCommitteeMember() { return StaffScope.COMMITTEE_ROLES.some((r) => this.hasRole(r)); }

  /** REQUEST breadth. Tenant equality is a precondition of every branch. */
  canSee(s) {
    if (!s || this.tenantId !== s.tenantId) return false;
    if (this.hasRole(StaffRole.INSTITUTION) || this.hasRole(StaffRole.OFFICE)) return true;
    if (this.hasRole(StaffRole.HOD) && this.department && this.department === s.department) return true;
    if (this.hasRole(StaffRole.FACULTY)) {
      for (const a of this.assignments) if (ClassKey.of(a).matches(s)) return true;
    }
    return false;
  }

  /** True iff this staff member personally teaches this subject to this class. */
  teachesSubject(c, subjectCode) {
    return this.assignments.some((a) => ClassKey.of(a).equals(c) && a.subjectCode === subjectCode);
  }

  /** True iff this staff member teaches at least one subject to this class. */
  teaches(c) { return this.assignments.some((a) => ClassKey.of(a).equals(c)); }

  /** The classes this staff member teaches — the ONLY values the class picker is built from. */
  classes() {
    const seen = new Map();
    for (const a of this.assignments) {
      const k = ClassKey.of(a);
      if (!seen.has(k.token())) seen.set(k.token(), k);
    }
    return [...seen.values()];
  }

  subjectsIn(c) { return this.assignments.filter((a) => ClassKey.of(a).equals(c)); }

  authors() { return this.assignments.length > 0; }
}
